### PR TITLE
Refactor collective communication send_partial, recv_partial, all_gather_partial C++ API

### PR DESCRIPTION
**Pull request by HermitSun**
### PR types

Others
### PR changes

APIs
### Describe

Refactor collective communication C++ API, including:
- send_partial
- recv_partial
- all_gather_partial

Try to fuse them into current APIs by adding `offset` & `numel` args.

### CODE REVIEW

File: paddle/fluid/distributed/collective/NCCLTools.cc [hunk 1]
```diff
@@ -44,5 +44,79 @@ std::string SerializeNCCLUniqueId(const ncclUniqueId& ncclID) {
...
+        reinterpret_cast<const bool*>(raw_pointer) + offset);
+  } else if (type == experimental::DataType::BFLOAT16) {
+    return reinterpret_cast<const void*>(
+        reinterpret_cast<const uint16_t*>(raw_pointer) + offset);
+  } else {
+    PADDLE_THROW(platform::errors::Unimplemented(
+        "Datatype %s in NCCL is not supported.", type));
+  }
+  return nullptr;
+}
```
Comment (LiYuRio) [inline paddle/fluid/distributed/collective/NCCLTools.cc]: 把这个逻辑删了，统一用slice吧，感觉没必要要自己实现

Comment (HermitSun) [reply paddle/fluid/distributed/collective/NCCLTools.cc]: done

File: paddle/fluid/distributed/collective/ProcessGroupNCCL.cc [hunk 2]
```diff
@@ -226,115 +244,88 @@ std::shared_ptr<ProcessGroup::Task> ProcessGroupNCCL::Broadcast(
...
+                                       stream);
+  };
+  // numel > 0 indicates the tensor need to be sliced
+  bool is_partial = numel > 0;
+  if (is_partial) {
+    phi::DenseTensor flattened_tensor;
+    flattened_tensor.ShareDataWith(*tensor);
+    flattened_tensor.Resize({tensor->numel()});
+    phi::DenseTensor partial_tensor =
+        flattened_tensor.Slice(offset, offset + numel);
```
Comment (LiYuRio) [inline paddle/fluid/distributed/collective/ProcessGroupNCCL.cc]: 能不能把这段抽出来，就不用总写了

Comment (HermitSun) [reply paddle/fluid/distributed/collective/ProcessGroupNCCL.cc]: done

**Review by LiYuRio — DISMISSED** [review]
LGTM

File: paddle/fluid/distributed/collective/ProcessGroupNCCL.cc [hunk 3]
```diff
@@ -226,48 +232,24 @@ std::shared_ptr<ProcessGroup::Task> ProcessGroupNCCL::Broadcast(
...
     bool sync_op,
     bool use_calc_stream) {
-  phi::DenseTensor tensor_flattened;
-  tensor_flattened.ShareDataWith(*tensor).Resize({tensor->numel()});
-  phi::DenseTensor tensor_recv =
-      tensor_flattened.Slice(offset, offset + length);
+  // numel > 0 indicates the tensor need to be sliced
+  if (numel > 0) {
+    phi::DenseTensor partial_tensor = GetPartialTensor(*tensor, offset, numel);
+    tensor = &partial_tensor;
```
Comment (LiYuRio) [inline paddle/fluid/distributed/collective/ProcessGroupNCCL.cc:245]: 这里的生命周期会有问题吧，partial_tensor出了这个作用域就被析构了

Comment (HermitSun) [reply paddle/fluid/distributed/collective/ProcessGroupNCCL.cc:245]: done

**Review by sljlp — APPROVED** [review]

**Pull request merged**